codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -185,125 +285,167 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
-            final Endpoint<Http2LocalFlowController> localEndpoint = http2Handler.connection().local();
-            streamId = localEndpoint.incrementAndGetNextStreamId();
-            try {
-                // Try to create a stream in OPEN state before writing headers, to catch errors on stream creation
-                // early on i.e. max concurrent streams limit reached, stream id exhaustion, etc.
-                localEndpoint.createStream(streamId, false);
-            } catch (Http2Exception e) {
-                promise.setFailure(e);
+            final DefaultHttp2FrameStream stream = (DefaultHttp2FrameStream) headersFrame.stream();
+            final Http2Connection connection = connection();
+            streamId = connection.local().incrementAndGetNextStreamId();
+            if (streamId < 0) {
+                promise.setFailure(new Http2NoMoreStreamIdsException());
                 return;
             }
-            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(streamId, headersFrame));
+            numBufferedStreams++;
+            // Set the stream id before completing the promise, as any listener added by a user will be executed
+            // before the below listener, and so the stream identifier is accessible in a user's listener.
+            stream.id(streamId);
+            // Ensure that the listener gets executed before any listeners a user might have attached.
+            // TODO(buchgr): Once Http2Stream2 and Http2Stream are merged this is no longer necessary.
+            writePromise = ctx.newPromise();
+            writePromise.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    numBufferedStreams--;
+
+                    Throwable cause = future.cause();
+                    Http2Stream connectionStream;
+                    if (cause == null && (connectionStream = connection.stream(streamId)) != null) {
+                        connectionStream.setProperty(streamKey, stream);
+                    } else {
+                        ctx.fireUserEventTriggered(Http2FrameStreamEvent.closed(stream));
+                    }
+                    if (cause == null) {
+                        promise.setSuccess();
+                    } else {
+                        promise.setFailure(cause);
+                    }
+                }
+            });
         }
-        http2Handler.encoder().writeHeaders(http2HandlerCtx, streamId, headersFrame.headers(),
-                                            headersFrame.padding(), headersFrame.isEndStream(), promise);
+        encoder().writeHeaders(ctx, streamId, headersFrame.headers(), headersFrame.padding(),
+                                            headersFrame.isEndStream(), writePromise);
     }
 
+    // TODO: Should we maybe also send Http2FrameStreamEvents for all the other stream state changes ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)